src/material-experimental/theming/_theming.scss

```diff
@@ -0,0 +1,142 @@
+@use 'sass:list';
+@use 'sass:map';
+@use 'sass:meta';
+@use 'sass:string';
+@use '@angular/material' as mat;
+
+// Whether to throw an error when a required dep is not configured. If false, the dep will be
+// automatically configured instead.
+$_error-on-missing-dep: false;
+
+// Applies the theme for the given component configuration.
+@mixin _apply-theme($tokens, $component) {
+  $id: map.get($component, id);
+  $tokens: map.deep-merge($tokens, map.get($component, customizations));
+
+  // NOTE: for now we use a hardcoded if-chain, but in the future when first-class mixins are
+  // supported, the configuration data will contain a reference to its own theme mixin.
+  @if $id == 'mat.card' {
+    @include mat.card-theme($tokens);
+  } @else if $id == 'mat.checkbox' {
+    @include mat.checkbox-theme($tokens);
+  } @else {
+    @error 'Unrecognized component theme: #{id}';
+  }
+}
+
+// Gets the transitive dependency configurations for the given list of component configurations.
+@function _get-transitive-deps($components, $configured: ()) {
+  // Mark the given components as configured.
+  @each $component in $components {
+    $configured: map.set($configured, map.get($component, id), true);
+  }
+  $new-deps: ();
+
+  // Check each of the given components for new deps.
+  @each $component in $components {
+    @each $dep-getter in mat.private-normalize-args-list(map.get($component, deps)) {
+      $dep: meta.call($dep-getter);
+      $dep-id: map.get($dep, id);
+      @if not (map.has-key($configured, $dep-id)) {
+        @if $_error-on-missing-dep {
+          @error 'Missing theme: `#{map.get($component, id)}` depends on `#{$dep-id}`.' +
+            ' Please configure the theme for `#{$dep-id}` in your call to `mat.theme`';
+        } @else {
+          $configured: map.set($configured, $dep-id, true);
```

Comment:
We should probably save all of the unconfigured deps for the end in case they're added out-of-order

Reply:
I think the order shouldn't matter. Missing deps are always configured with an empty customization map, so it doesn't really matter in what order the empty map is merged into the combined map